src/Mod/Sketcher/Gui/Command.cpp

```diff
@@ -1213,6 +1191,250 @@ bool CmdSketcherGrid::isActive()
+    CmdSketcherSnap& operator= (const CmdSketcherSnap&) = delete;
+    CmdSketcherSnap& operator= (CmdSketcherSnap&&) = delete;
+};
+
+CmdSketcherSnap::CmdSketcherSnap()
+    : Command("Sketcher_Snap")
+{
+    sAppModule = "Sketcher";
+    sGroup = "Sketcher";
+    sMenuText = QT_TR_NOOP("Toggle Snap");
+    sToolTipText = QT_TR_NOOP("Toggle all snapping functionalities. In the menu you can toggle individually 'Snap to Grid', 'Snap to Objects' and further snap settings");
+    sWhatsThis = "Sketcher_Snap";
+    sStatusTip = sToolTipText;
+    eType = 0;
+}
+
+void CmdSketcherSnap::updateIcon(bool value)
+{
+    static QIcon active = Gui::BitmapFactory().iconFromTheme("Sketcher_Snap");
+    static QIcon inactive = Gui::BitmapFactory().iconFromTheme("Sketcher_Snap_Deactivated");
+
+    auto* pcAction = qobject_cast<Gui::ActionGroup*>(getAction());
+    pcAction->setIcon(value ? active : inactive);
+}
+
+void CmdSketcherSnap::activated(int iMsg)
+{
+    Q_UNUSED(iMsg);
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+    bool value = !hGrp->GetBool("Snap", true);
+    
+    hGrp->SetBool("Snap", value);
+
+    updateIcon(value);
+
+    //Update the widget :
+    if (!_pcAction)
+        return;
+
+    Gui::ActionGroup* pcAction = qobject_cast<Gui::ActionGroup*>(_pcAction);
+    QList<QAction*> a = pcAction->actions();
+
+    auto* ssa = static_cast<SnapSpaceAction*>(a[0]);
+    ssa->updateWidget();
+}
+
+Gui::Action* CmdSketcherSnap::createAction()
+{
+    auto* pcAction = new Gui::ActionGroup(this, Gui::getMainWindow());
+    pcAction->setDropDownMenu(true);
+    pcAction->setExclusive(false);
+    applyCommandData(this->className(), pcAction);
+
+    SnapSpaceAction* ssa = new SnapSpaceAction(pcAction);
+    pcAction->addAction(ssa);
+
+    _pcAction = pcAction;
+
+    QObject::connect(pcAction, &Gui::ActionGroup::aboutToShow, [ssa](QMenu* menu) {
+        Q_UNUSED(menu)
+            ssa->updateWidget();
+    });
+
+    // set the right pixmap
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+    updateIcon(hGrp->GetBool("Snap", true));
+
+    return pcAction;
+}
+
+void CmdSketcherSnap::languageChange()
+{
+    Command::languageChange();
+
+    if (!_pcAction)
+        return;
+
+    Gui::ActionGroup* pcAction = qobject_cast<Gui::ActionGroup*>(_pcAction);
+    QList<QAction*> a = pcAction->actions();
+
+    auto* ssa = static_cast<SnapSpaceAction*>(a[0]);
+    ssa->languageChange();
+}
+
+bool CmdSketcherSnap::isActive()
+{
+    auto* vp = getInactiveHandlerEditModeSketchViewProvider();
+
+    if (vp) {
+        ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+        bool value = hGrp->GetBool("Snap", true);
+
+        updateIcon(value);

```

Comment:
And for setting the icon!!

Reply:
I didn't knew GetBool / SetBool were heavy so I didn't paid attention not to use them everywhere.
So what should I do? Because we still need to register the parameters.
Maybe I can check the parameter only on 'CreateAction' and register the state in a bool for activated and isActive?

Reply:
Take a look to my commits how I approached it. I do not mean it is perfect, but it is targeting a reduction of signaling overhead. Specially in the isActive function. By making the state of the button a data member and implementing an observer, the data member is always up-to-date without need for polling. Then isActive only needs to check a data member variable. 

Another part of if is passing the state as a parameter to updateWidget, this way the widgetaction does not need to ask for it again.